### PR TITLE
Avoid kernel launch when number of threads is 0

### PR DESCRIPTION
**Pull request by ShreyasKhandekar**
Resolves https://github.com/chapel-lang/chapel/issues/23658 and adds a test for it
Resolves https://github.com/chapel-lang/chapel/issues/20635 and updates the associated test
Resolves https://github.com/chapel-lang/chapel/issues/23095

This changes the behavior of GPU diagnostics as well to not register this as a kernel launch.
We also print a message informing the developer that a kernel was not launched when using `--debugGpu`.

Pending Nvidia and AMD testing for latest commit.

### CODE REVIEW

**Comment by e-kayrakli** [comment]
This could also potentially solve:

- https://github.com/chapel-lang/chapel/issues/20635
- https://github.com/chapel-lang/chapel/issues/23095

@ShreyasKhandekar -- I assigned you to both of these issues with that hope. Could you check whether this PR resolves them both before/after merging and close them if/when they're resolved. If they remain problematic, please feel free to un-assign yourself.

p.s. A PR that resolves 3 issues at the same time doesn't happen every week! :)

File: runtime/src/chpl-gpu.c [hunk 1]
```diff
@@ -285,25 +285,28 @@ inline void chpl_gpu_launch_kernel_flat(int ln, int32_t fn,
   va_list args;
   va_start(args, nargs);
 
-  chpl_gpu_diags_verbose_launch(ln, fn, chpl_task_getRequestedSubloc(),
-      blk_dim, 1, 1);
-  chpl_gpu_diags_incr(kernel_launch);
+  if (num_threads > 0){
+    CHPL_GPU_DEBUG("No kernel launched since num_threads is 0\n");    )
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:289]: Shouldn't this be in an `else`?

File: runtime/src/chpl-gpu.c [hunk 2]
```diff
@@ -285,25 +285,28 @@ inline void chpl_gpu_launch_kernel_flat(int ln, int32_t fn,
...
+        blk_dim, 1, 1);
+    chpl_gpu_diags_incr(kernel_launch);
 
-  chpl_gpu_impl_launch_kernel_flat(ln, fn,
-                                   name,
-                                   num_threads, blk_dim,
-                                   stream,
-                                   nargs, args);
+    chpl_gpu_impl_launch_kernel_flat(ln, fn,
+                                    name,
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:296]: weird indentation. `ln`, `name` etc should start in the same column

File: test/gpu/native/basics/bug23658.chpl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// TBD
+use GpuDiagnostics;
+
+config const gpuDiags = true;
+const sz = 10;
+
+if gpuDiags then startGpuDiagnostics();
+
+on here.gpus[0] {
+  const s = sz;
```
Comment (e-kayrakli) [inline test/gpu/native/basics/bug23658.chpl:12]: You don't really need this. You can use `sz` as the size of the array.

File: test/gpu/native/basics/bug23658.chpl [hunk 4]
```diff
@@ -0,0 +1,24 @@
+// this code is from https://github.com/chapel-lang/chapel/issues/23658 verbatim
```
Comment (e-kayrakli) [inline test/gpu/native/basics/bug23658.chpl:1]: Doesn't look very "verbatim" :) You can just drop that word

File: test/gpu/native/basics/bug23658.chpl [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  var Arr: [0..<s] int;
+  foreach i in 1..0 do {
+    Arr[i] = i;
+  }
+  writeln(Arr);
+}
+
+if gpuDiags {
+  stopGpuDiagnostics();
+  assertGpuDiags(kernel_launch=1, host_to_device=0, device_to_host=sz,
```
Comment (e-kayrakli) [inline test/gpu/native/basics/bug23658.chpl:22]: This 1 launch is coming from array initialization. I expect that you will _not_ get any launches with `unified_memory`. You can drop `kernel_launch` argument, and use `kernel_launch_aod` and `kernel_launch_um` args separately to set different expected values for those cases.

(This makes me think we should stop doing separate things and have `unified_memory` do GPU-based initialization)

File: runtime/src/chpl-gpu.c (hunk 1 above)
Comment (e-kayrakli) [reply runtime/src/chpl-gpu.c:289]: Please make sure you test this with your test case to see it fires correctly. (You can see this output if you run your executable with `--debugGpu`)

**Review by e-kayrakli — COMMENTED** [review]
I have some nitpicky comments inline. Two main ones that I believe are correctness issues:

- `CHPL_GPU_DEBUG` output seems to be in the wrong place
- `assertGpuDiags` in the test will probably trigger with `unified_memory`

I am also curious whether you got a chance to try this with other issues this can potentially fix. Looking at those issues, I don't see there's any interesting snippet that's sufficiently different than what you have here. So, if they are resolved, I don't think you need to add more tests for those issues. But you can edit your PR messages to add "Resolves 12345" for those two, too.

File: runtime/src/chpl-gpu.c (hunk 1 above)
Comment (ShreyasKhandekar) [reply runtime/src/chpl-gpu.c:289]: Should I have the test use the `--debugGpu` flag? Do you know if I need to sanitize the output using a prediff if I'm using that flag?

Comment (e-kayrakli) [reply runtime/src/chpl-gpu.c:289]: Yes. I think the best way to do that is just to grep this exact output message and have that be the sole thing in the good file. You should avoid locking the rest of the debugGpu output. It is not something we test/lock as a whole (I like that we don't, I want to be able to change it easily), but for specific things it looks like we have some tests that use it in execopts already.

BTW The output should say `<=0`, or anything that's not `is 0`.

**Pull request merged**